File: apple/Sources/Plugins/JaEchoTestHandle.swift
//
//  Dummy.swift
//
//
//  Created by Hamza Jadid on 16/09/2024.
//

import Foundation
import UniFFI

/// EchoTest plugin handle,
///
/// The purpose of this plugin is for testing. A peer attaching to this plugin will receive the same packets he
/// sends.
public final class JaEchoTestHandle {
    let handle: EchotestHandle
    private var continuation: AsyncStream<JaEchoTestEvent>.Continuation?

    /// Get an async stream of incoming Janus echotest events, check ``JaEchotestEvent``
    ///
    /// - Returns: An async stream of incoming events as ``JaEchotestEvent``
    public var events: AsyncStream<JaEchoTestEvent> {
        get async {
            await handle.startEventLoop(cb: self)

            return AsyncStream { continuation in
                self.continuation = continuation
            }
        }
    }

    init(handle: EchotestHandle) {
        self.handle = handle
    }

    /// Start the testing
    ///
    /// - Parameters:
    ///     - audio: enable/disable sending back audio
    ///     - video: enable/disable sending back video
    ///     - bitrate: to cap bitrate at the provided value
    public func start(
        audio: Bool = false,
        video: Bool = false,
        bitrate: UInt32? = nil
    ) async throws {
        try await handle.start(
            audio: audio,
            video: video,
            bitrate: bitrate
        )
    }

    /// Start the testing
    ///
    /// - Parameters:
    ///     - audio: enable/disable sending back audio
    ///     - video: enable/disable sending back video
    ///     - bitrate: to cap bitrate at the provided value
    ///     - jsep: JavaScript establishment protocol
    ///     - timeout: The maximum amount of time to wait on an acknowledgment before we consider
    ///     the request as failed or times out
    public func start(
        audio: Bool = false,
        video: Bool = false,
        bitrate: UInt32? = nil,
        jsep: Jsep,
        timeout: TimeInterval
    ) async throws {
        try await handle.startWithJsep(
            audio: audio,
            video: video,
            bitrate: bitrate,
            jsep: jsep,
            timeout: timeout
        )
    }
}

/// Asynchronous incoming event type
public enum JaEchoTestEvent {
    case result(echotest: String, result: String)
    case resultWithJsep(echotest: String, result: String, jsep: Jsep)
}

extension JaEchoTestHandle: EchotestHandleCallback {
    public func onResult(echotest: String, result: String) {
        continuation?.yield(.result(echotest: echotest, result: result))
    }

    public func onResultWithJsep(echotest: String, result: String, jsep: Jsep) {
        continuation?.yield(
            .resultWithJsep(echotest: echotest, result: result, jsep: jsep)
        )
    }
}

File: android/plugins/src/main/java/com/ghamza/janus/plugins/JaEchotestHandle.kt
package com.ghamza.janus.plugins

import com.ghamza.janus.bindings.EchotestHandle
import com.ghamza.janus.bindings.EchotestHandleCallback
import com.ghamza.janus.bindings.Jsep
import kotlinx.coroutines.channels.ProducerScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.callbackFlow
import java.time.Duration

class JaEchotestHandle(val handle: EchotestHandle): EchotestHandleCallback {
    private var events: ProducerScope<JaEchotestEvent>? = null

    fun stream() = callbackFlow {
        events = this
        handle.startEventLoop(this@JaEchotestHandle)
        awaitClose { }
    }

    suspend fun start(audio: Boolean = false, video: Boolean = false, bitrate: UInt? = null) {
        handle.start(audio = audio, video = video, bitrate = bitrate)
    }

    suspend fun start(
        audio: Boolean = false,
        video: Boolean = false,
        bitrate: UInt? = null,
        jsep: Jsep,
        timeout: Duration
    ) {
        handle.startWithJsep(audio = audio, video = video, bitrate = bitrate, jsep = jsep, timeout = timeout)
    }

    override fun onResult(echotest: String, result: String) {
        events?.trySend(JaEchotestEvent.Result(echotest = echotest, result = result))
    }

    override fun onResultWithJsep(echotest: String, result: String, jsep: Jsep) {
        events?.trySend(JaEchotestEvent.ResultWithJsep(echotest = echotest, result = result, jsep = jsep))
    }
}

sealed interface JaEchotestEvent {
    data class Result(val echotest: String, val result: String): JaEchotestEvent
    data class ResultWithJsep(val echotest: String, val result: String, val jsep: Jsep): JaEchotestEvent
}
